### PR TITLE
build(protocol-definitions): use node16 module resolution

### PR DESCRIPTION
**Pull request by DLehenbauer**
Generated code for type tests needs to add a .js extension:
```
[build:esnext] src/test/types/validateProtocolDefinitionsPrevious.generated.ts:10:31 - error TS2835: Relative import paths need explicit file extensions in EcmaScript imports when '--moduleResolution' is 'node16' or 'nodenext'. Did you mean '../../index.js'?
[build:esnext] 
[build:esnext] 10 import type * as current from "../../index";
[build:esnext]                                  ~~~~~~~~~~~~~
[build:esnext] 
[build:esnext] 
[build:esnext] Found 1 error in src/test/types/validateProtocolDefinitionsPrevious.generated.ts:10
[build:esnext] 
[build:esnext] npm run build:esnext exited with code 2
```

Semi-automatic PR to switch to node16 module resolution

### CODE REVIEW

File: common/lib/protocol-definitions/package.json [hunk 1]
```diff
@@ -11,11 +11,52 @@
...
+		"./beta": {
+			"import": {
+				"types": "./lib/protocol-definitions-beta.d.ts",
+				"default": "./lib/index.js"
+			},
+			"require": {
+				"types": "./dist/protocol-definitions-beta.d.ts",
+				"default": "./dist/index.js"
+			}
+		},
```
Comment (jason-ha) [inline common/lib/protocol-definitions/package.json:45]: No beta, right?

File: common/lib/protocol-definitions/package.json [hunk 2]
```diff
@@ -11,11 +11,52 @@
...
+		"./public": {
+			"import": {
+				"types": "./lib/protocol-definitions-public.d.ts",
+				"default": "./lib/index.js"
+			},
+			"require": {
+				"types": "./dist/protocol-definitions-public.d.ts",
+				"default": "./dist/index.js"
+			}
+		},
```
Comment (jason-ha) [inline common/lib/protocol-definitions/package.json:35]: I would skip the ./public phase for this and just trim.

**Review by jason-ha — CHANGES_REQUESTED** [review]

**Comment by jason-ha** [comment]
Looks like build tools need updated as well for the type test generation.

File: common/lib/protocol-definitions/package.json (hunk 2 above)
Comment (jason-ha) [reply common/lib/protocol-definitions/package.json]: exactly

File: common/lib/protocol-definitions/package.json [hunk 3]
```diff
@@ -66,12 +87,19 @@
 		"eslint-config-prettier": "~9.0.0",
 		"prettier": "~3.0.3",
 		"rimraf": "^2.6.2",
-		"tsc-multi": "^1.1.0",
 		"typescript": "~5.1.6"
 	},
 	"packageManager": "pnpm@7.32.3+sha512.c28d73f0d82a6c4712a40b91aa43d7397e119b16f35e5e75fe9b7e8fd4f2fc6dfbb68bb3ffac3473a958bbafa7888b79ec384ad122537378089a3a2a19b9900e",
```
Comment (jason-ha) [inline common/lib/protocol-definitions/package.json:92]: @tylerbutler, not for now - this should be updated, yeah?

Comment (tylerbutler) [reply common/lib/protocol-definitions/package.json:92]: Yeah, independent packages tend to get missed when people do version updates.

**Review by jason-ha — APPROVED** [review]

File: common/lib/protocol-definitions/package.json (hunk 2 above)
Comment (jason-ha) [reply common/lib/protocol-definitions/package.json]: (That was my ask. `exactly` was a poor reply - I think I read Tyler's "or" as "and".)

File: common/lib/protocol-definitions/package.json [hunk 4]
```diff
@@ -24,15 +45,15 @@
...
+		"api": "fluid-build . --task api",
+		"api-extractor:commonjs": "api-extractor run --config ./api-extractor-cjs.json",
+		"api-extractor:esnext": "api-extractor run --local",
+		"build": "fluid-build . --task build",
+		"build:compile": "fluid-build . --task compile",
+		"build:docs": "fluid-build . --task api",
+		"build:esnext": "tsc --project ./tsconfig.json",
 		"check:are-the-types-wrong": "attw --pack . --entrypoints .",
 		"check:release-tags": "api-extractor run --local --config ./api-extractor-lint.json",
 		"ci:build": "npm run build:compile",
```
Comment (jason-ha) [inline common/lib/protocol-definitions/package.json:59]: `ci:build` only calls `build:compile` but now relies on `api` to generate the rollup files. The lint step will fail without `api` being run. I think `ci:build` should be `fluid-build --task ci:build`.

**Review by jason-ha — CHANGES_REQUESTED** [review]

File: common/lib/protocol-definitions/package.json (hunk 4 above)
Comment (tylerbutler) [reply common/lib/protocol-definitions/package.json:59]: ci:build doesn't need the api files, but linting does. Thus I changed the lint task to use fluid-build, but left ci:build calling compile only. Lint is called independently in CI, so when it's run, it'll do the api-extractor stuff too.

**Review by jason-ha — APPROVED** [review]

File: common/lib/protocol-definitions/package.json (hunk 4 above)
Comment (jason-ha) [reply common/lib/protocol-definitions/package.json:59]: I approved, but I don't think that is correct. Build and ci:build should produce everything required to publish package. Linting should not generate a portion of the build. There should be a follow up to correct that.

**Pull request merged**